File: backend/app/schemas/ai.py
from typing import Literal

from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    # Assistant turns that only carried tool calls have null content.
    content: str | None = Field(default=None, max_length=32_000)


class ChatRequest(BaseModel):
    messages: list[ChatMessage] = Field(min_length=1, max_length=50)


class Proposal(BaseModel):
    action: Literal["create", "edit", "delete"]
    section: str
    fields: dict | None = None        # create/edit drafted fields
    record_id: str | None = None      # edit/delete target
    before: dict | None = None        # edit only: current values
    warnings: list[str] = Field(default_factory=list)


class ChatResponse(BaseModel):
    answer: str
    tools_used: list[str]
    proposals: list[Proposal] = Field(default_factory=list)


class AiStatus(BaseModel):
    """Minimal, viewer-safe AI status for the chat panel. Deliberately omits
    base_url so the local LLM endpoint is never exposed to non-admins."""
    enabled: bool
    model: str | None

File: frontend/src/api/ai.ts
import { csrfHeader } from "./csrf";

export interface ChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface Proposal {
  action: "create" | "edit" | "delete";
  section: string;
  fields?: Record<string, unknown> | null;
  record_id?: string | null;
  before?: Record<string, unknown> | null;
  warnings?: string[];
}

export interface ChatResponse {
  answer: string;
  tools_used: string[];
  proposals?: Proposal[];
}

export class AiUnavailableError extends Error {}

// 10-minute timeout matches the backend's httpx timeout — needed for CPU-only
// Ollama cold starts which can take several minutes to load the model.
const CHAT_TIMEOUT_MS = 600_000;

export async function sendChat(messages: ChatMessage[]): Promise<ChatResponse> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), CHAT_TIMEOUT_MS);
  try {
    const res = await fetch("/api/ai/chat", {
      method: "POST",
      headers: { "Content-Type": "application/json", ...csrfHeader() },
      body: JSON.stringify({ messages }),
      signal: controller.signal,
    });
    if (res.status === 503) throw new AiUnavailableError("AI assistant is unavailable.");
    if (!res.ok) throw new Error("Chat request failed");
    return res.json();
  } finally {
    clearTimeout(timer);
  }
}
